### PR TITLE
Hello, a few thougts on some rust code..

### PR DESCRIPTION
**Pull request by alphastrata**
TLDR : big changes to `inpaint.rs`, check em out if you're interested in rustying up your code, the rest is `cargo fmt`.

Hello, you don't know me, but, I happen to write rust at my day-job, love space, love to see more astro-crates out there and after taking a look at some of it... had thoughts!

I'm not at all an imaging or algo expert, but I've just taken a quick look at one of your source files and seen a lot of room for further oxidisation.

They're just ideas on how to make things a little rustier, hopefully they lend insights to you that are worthwhile both stylistically and, at runtime.

Happy to see them outright rejected if you're not interested etc.

However, if you are, a couple of things:

- use clippy `cargo clippy --no-deps` it will lint your code, it is one of the ecosystem's most bestest gooderest badass dtools!
- use `cargo fmt` it would be extremely difficult to have folk contribute to the project without it. (get your IDE to do it for you on save!)
- use iterators, you'll find that (should you feel game enough to inspect the assembly) it creates significant performance optimisations to your code, you'll also find that a significant enough proportion of rust is written in this more functional style.
- embrace `match`, it's very good, far superior to huge-ass `if-else` chaining and! it a lot of cases it's more flexible, including its uses in enforcing particular system states via `enum` destructuring!  https://cheats.rs/#pattern-matching for a good reference.

- I noticed comments about some 'slow' code, but didn't see any evidence of benchmarks, I can see you've explored the test harness that ships with rust, the benchmark ones are also best-in-class, I'd suggest crates like `criterion`, `iai` and apps like `, and, of course `cargo flamegraph` -- if you're on a linux-esk system these will all be trivial to setup.

- embrace the type system! a lot of the `if x > 0` checks can be done away with, it also looks as though you've made a conspicuous decision to use i/u/f32 values, if you're targeting a cpu from this decade just use `usize/isize`. If you're particularly worried about the way a `struct` will be packed into memory, you can get C-style packing with `repr(c)` https://doc.rust-lang.org/nomicon/other-reprs.html, there are other options. Rust natively will give you bugger-all in terms of layout guarantees -- which is the only reason I can think you'd try this... (right now at least)

- Don't make so many allocations, we have the `mut` keyword which is well and good, however usually when instantiating collections you'll see them built in memory with `Iterator`s rather than explicitly assigned to things. 

A few final notes:
- I've used `.map()` and `.collect()` quite a bit, this is pretty par-for-the-course in most rust libraries, so if you're going to continue using rust, expect to see more of this & expect to consume APIs in this way (Rayon, for example flat-out doesn't work with the for-loop syntax [which is syntactic sugar anyway.. but you can google that if you're interested]).
If you've not used a lot of Iterators before, this can seem weird at first, but I'd encourage you to persevere! they're all incredibly powerful, and, as rust is such an extensive language these days you'll get wonderful methods that others have already implemented for you!.

Lastly, `.map()` calls are lazy, many of rust's iterators are -- worth keeping this in mind.

Really, really lastly, please accept my sincerest apologies if any of this is coming across as condescending etc, that's not the intention, I don't know you, or, how much development experience you do/don't have etc.

-J

### CODE REVIEW

**Comment by kmgill** [comment]
Thank you so much! I am just learning Rust on my own and it's a huge relief to get some expert eyes on the code to catch me on any bad practices I may have developed. I came at Rust with a very Java/Python-heavy mindset and was quickly (and somewhat painfully) dispelled of that idea, and it's been a slow journey since then.  

I am going to review all your changes and suggestions in the coming days and will likely work them into the Rust projects that use this crate as a dependency.

**Comment by alphastrata** [comment]
> Thank you so much! I am just learning Rust on my own and it's a huge relief to get some expert eyes on the code to catch me on any bad practices I may have developed. I came at Rust with a very Java/Python-heavy mindset and was quickly (and somewhat painfully) dispelled of that idea, and it's been a slow journey since then.
> 
> I am going to review all your changes and suggestions in the coming days and will likely work them into the Rust projects that use this crate as a dependency.

Happy to see this sort of crate content grow, lemme know if you have followups/if anything 'should' work elsewhere but is a struggle to get working.

Happy rusting :)

**Comment by alphastrata** [comment]
@kmgill you cool with me putting out a bit of a call to arms for your repos? (i'd very, very much like for rust to succeed in this space.. (the space space that is haha!)
